opencog/cogserver/server/AgentRunnerBase.cc

```diff
@@ -0,0 +1,165 @@
+}
+
+unsigned long AgentRunnerBase::getCycleCount() const
+{
+    return cycleCount;
+}
+
+void AgentRunnerBase::addAgent(AgentPtr a)
+{
+    agents.push_back(a);
+}
+
+void AgentRunnerBase::removeAgent(AgentPtr a)
+{
+    AgentSeq::iterator ai = std::find(agents.begin(), agents.end(), a);
+    if (ai != agents.end()) {
+        agents.erase(ai);
+        cogserver().systemActivityTable().clearActivity(a);
+        a->stop();
+        logger().debug("[CogServer::%s] stopped agent \"%s\"", name.c_str(),
+            a->to_string().c_str());
+    }
+}
+
+void AgentRunnerBase::removeAllAgents(const std::string& id)
+{
+    // place agents with classinfo().id == id at the end of the container
+    AgentSeq::iterator last =
+        std::partition(agents.begin(), agents.end(),
+                       boost::bind(equal_to_id(), _1, id));
+
+    std::for_each(last, agents.end(), [] (AgentPtr &a) {
+        cogserver().systemActivityTable().clearActivity(a);
+        a->stop();
+    });
+
+    // remove those agents from the main container
+    agents.erase(last, agents.end());
+
+    logger().debug("[CogServer::%s] stopped all agents of type \"%s\"",
+        name.c_str(), id.c_str());
+}
+
+void AgentRunnerBase::removeAllAgents()
+{
+    for (auto &a: agents)
+    {
+        cogserver().systemActivityTable().clearActivity(a);
+        a->stop();
+    }
+    agents.clear();
+    logger().debug("[CogServer::%s] stopped all agents", name.c_str());
+}
+
+void AgentRunnerBase::runAgent(AgentPtr a)
+{
+    size_t mem_start = getMemUsage();
+    int atoms_start = atomspace().get_size();
+
+    logger().debug("[CogServer::%s] begin to run mind agent: %s, [cycle = %d]",
+                   name.c_str(), a->classinfo().id.c_str(), cycleCount);
+
+    auto timer_start = system_clock::now();
+
+    a->resetUtilizedHandleSets();
+    a->run();
+
+    auto timer_end = system_clock::now();
+
+    size_t mem_end = getMemUsage();
+    int atoms_end = atomspace().get_size();
+
+    size_t mem_used;
+    if (mem_start > mem_end)
+        mem_used = 0;
+    else
+        mem_used = mem_end - mem_start;
+
+    int atoms_used;
+    if (atoms_start > atoms_end)
+        atoms_used = 0;
+    else
+        atoms_used = atoms_end - atoms_start;
+
+    auto elapsed = timer_end - timer_start;
+    auto elapsed_secs = duration_cast<duration<float>>(elapsed).count();
+    logger().debug("[CogServer::%s] running mind agent: %s, elapsed time (sec): "
+            "%f, memory used: %d, atom used: %d [cycle = %d]", name.c_str(),
+            a->classinfo().id.c_str(), elapsed_secs, mem_used, atoms_used,
+            cycleCount);
+
+    cogserver().systemActivityTable().logActivity(a, elapsed, mem_used,
+        atoms_used);
+}
+
+void SimpleRunner::processAgents()
+{
+    AgentSeq::const_iterator it;
+    for (it = agents.begin(); it != agents.end(); ++it) {
+        AgentPtr agent = *it;
+        if ((cycleCount % agent->frequency()) == 0)
+            runAgent(agent);
```

Comment:
Yes, this "SimpleRunner" (Yes, the name is not appropriate, I was looking for an alternative name too. And, your suggestion seems a good one) is the cooperative time-sliced runner which behaves like before and runs in the main thread. It is the compatibility interface.